### DISCUSSION TITLE
How to read nextjs server action function name and parameters in middleware?

### DISCUSSION THREAD

**Discussion by amplicity** [vercel/next.js#68561] [App Router] [answered]
When [server actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) are called, they manifest as a POST request from the user's current route. I'm trying to get information about the function name and params in [middleware](https://nextjs.org/docs/pages/building-your-application/routing/middleware)
```
export default async function middleware(request: NextRequest) {
  console.log('request', request)
}
```
When I inspect request, I find no references to the function name or the params passed, but i'm only crudely inspecting and haven't ruled out serialization. Is there a way to view the server action's function name and parameters that called the POST route?

**Comment by Juneezee** [comment, marked as answer]
I'm afraid no. 

Next.js identifies the server action through the `Next-Action` header. You can see a demonstration in this Youtube video: https://youtu.be/mwNZ34sxIGw?t=293

**Reply by amplicity** [reply]
Thank you, have been looking for a reference like you shared.

**Reply by amplicity** [reply]
Do you know of any way to reference/decode that id to something meaningful?

**Reply by Juneezee** [reply]
> Do you know of any way to reference/decode that id to something meaningful?

I'm sorry to disappoint you. All I know is that Next.js generates a unique ID for each Server Action on build time. I didn't delve in Next.js source code to understand how the ID works internally.

Also, Next.js _could_ be adding salt in the future so rebuilds won't produce the same IDs anymore, see https://github.com/vercel/next.js/issues/63804#issuecomment-2025660356. 

> We're also adding hash salt rotation to production build, so rebuilds will be protected in this case and all ids will change.

So unless Next.js is willing to provide us a way to identify the Server Action through the `Next-Action` header, I don't see a reliable way of doing it using hacks.
